File: python/examples/vtk/isobands.py
"""
Interactive isoband extraction example with VTK

The scalar field is computed as signed distance from a random plane.
Multiple isobands are extracted simultaneously with every other band selected.
Boundary curves between bands are visualized as red tubes.
Extraction time is averaged over the last 100 frames.
"""

import vtk
import numpy as np
import trueform as tf

# Import utilities
from util import (
    MeshData,
    numpy_to_polydata,
    compute_centering_and_scaling_transform,
    curves_to_polydata,
    BaseInteractor,
    RollingAverage,
    format_time_ms,
    create_text_actor,
    create_parser,
)

# Set to True to test dynamic mesh (OffsetBlockedArray)
USE_DYNAMIC_MESH = True


class IsobandInteractor(BaseInteractor):
    """Interactor for isoband visualization with scrolling and randomization"""

    def __init__(self, mesh_data, isobands_polydata, curve_polydata, text_actor):
        super().__init__()

        # Keep mesh data (wraps trueform mesh + actor)
        self.mesh_data = mesh_data
        self.mesh = mesh_data.mesh  # Convenience reference
        self.isobands_polydata = isobands_polydata
        self.curve_polydata = curve_polydata
        self.text_actor = text_actor

        # Scalar field data
        self.scalars = None
        self.distance = 0.0  # Offset for isoband levels
        self.min_d = 0.0     # Min scalar value
        self.max_d = 0.0     # Max scalar value
        self.spacing = 0.1   # Will be computed in reset_plane

        # Timing
        self.times = RollingAverage(maxlen=100)

        # Initialize random plane (but don't compute bands yet - interactor not attached)
        self.reset_plane()

        # Add custom event handlers for scrolling and keys
        self.AddObserver("KeyPressEvent", self.on_key_press)
        self.AddObserver("MouseWheelForwardEvent", self.on_mouse_wheel_forward)
        self.AddObserver("MouseWheelBackwardEvent", self.on_mouse_wheel_backward)

    def reset_plane(self):
        """Create a new random plane and compute scalar field"""
        # Create random plane through a random point
        random_normal = np.random.randn(3).astype(self.mesh.dtype)
        random_normal /= np.linalg.norm(random_normal)
        random_point = self.mesh.points[np.random.randint(0, len(self.mesh.points))]

        plane = tf.Plane.from_point_normal(random_point, random_normal)

        # Compute scalar field (signed distance to plane)
        self.scalars = tf.distance_field(self.mesh.points, plane)
        self.distance = 0.0

        # Cache min/max for adaptive level spacing
        self.min_d = float(np.min(self.scalars))
        self.max_d = float(np.max(self.scalars))
        self.spacing = (self.max_d - self.min_d) / 10.0  # N = 10 bands

    def compute_bands(self):
        """Extract isobands with curves and update visualization"""
        import time as time_module

        # Compute adaptive threshold levels (same algorithm as isobands.cpp)
        N = 10  # Number of bands

        # Use fmod to wrap offset within spacing (infinite scrolling)
        wrapped_offset = self.distance % self.spacing
        if wrapped_offset < 0:
            wrapped_offset += self.spacing

        # Generate evenly spaced cut values offset by wrapped_offset
        # Extend below and above the range so wrapping is seamless
        cutvalues = np.array(
            [self.min_d + wrapped_offset + i * self.spacing for i in range(-1, N + 2)],
            dtype=self.mesh.dtype
        )

        # Select every other band (alternating pattern)
        # Now we have N+2 bands (indices 0 to N+1)
        parity = int(np.floor(self.distance / self.spacing)) & 1
        selected_bands = np.array([i for i in range(N + 2) if (i & 1) == parity], dtype=np.int32)

        # Extract isobands with curves and timing
        start_time = time_module.perf_counter()
        (band_faces, band_points), labels, (paths, curve_points) = tf.isobands(
            self.mesh, self.scalars, cutvalues, selected_bands, return_curves=True
        )
        elapsed = time_module.perf_counter() - start_time

        # Update timing
        self.times.add(elapsed)
        avg_time = self.times.get_average()
        self.text_actor.SetInput(f"Isobands time: {format_time_ms(avg_time)}")

        # Convert isobands mesh to polydata
        if len(band_faces) > 0 and len(band_points) > 0:
            poly = numpy_to_polydata(band_points, band_faces)
            self.isobands_polydata.ShallowCopy(poly)
        else:
            # No bands - clear visualization
            self.isobands_polydata.Initialize()

        self.isobands_polydata.Modified()

        # Convert curves to polydata
        if len(paths) > 0 and len(curve_points) > 0:
            curve_poly = curves_to_polydata(paths, curve_points)
            self.curve_polydata.ShallowCopy(curve_poly)
        else:
            # No curves - clear visualization
            self.curve_polydata.Initialize()

        self.curve_polydata.Modified()

        # Only render if interactor is attached
        if self.GetInteractor() is not None:
            self.GetInteractor().Render()

    def on_key_press(self, obj, event):
        """Handle key press events"""
        key = self.GetInteractor().GetKeySym()
        if key == "n":
            # Randomize plane
            self.reset_plane()
            self.compute_bands()
        else:
            # Pass to base class for camera controls
            vtk.vtkInteractorStyleTrackballCamera.OnKeyPress(self)

    def on_mouse_wheel_forward(self, obj, event):
        """Handle mouse wheel forward"""
        if self.GetInteractor().GetControlKey():
            # Move isobands forward (same as C++: spacing * 0.1)
            self.distance += self.spacing * 0.1
            self.compute_bands()
        else:
            # Pass to base class for zoom
            vtk.vtkInteractorStyleTrackballCamera.OnMouseWheelForward(self)

    def on_mouse_wheel_backward(self, obj, event):
        """Handle mouse wheel backward"""
        if self.GetInteractor().GetControlKey():
            # Move isobands backward (same as C++: spacing * 0.1)
            self.distance -= self.spacing * 0.1
            self.compute_bands()
        else:
            # Pass to base class for zoom
            vtk.vtkInteractorStyleTrackballCamera.OnMouseWheelBackward(self)


def main():
    # Parse command line arguments
    parser = create_parser("Interactive isoband extraction", mesh_args=1)
    parser.epilog = """
Controls:
  N              Randomize cutting plane
  Ctrl + Scroll  Move isoband levels
  Mouse drag     Rotate camera
"""
    args = parser.parse_args()
    mesh_file = args.mesh

    # Load mesh
    faces, points = tf.read_stl(mesh_file)

    # Optionally convert to dynamic mesh
    if USE_DYNAMIC_MESH:
        faces = tf.as_offset_blocked(faces)

    # Center and scale mesh
    transform = compute_centering_and_scaling_transform(points, target_radius=10.0)

    # Apply transformation
    points_homogeneous = np.hstack([points, np.ones((len(points), 1), dtype=points.dtype)])
    points_transformed = (transform @ points_homogeneous.T).T[:, :3]

    # Create trueform mesh FIRST (this is the primary data structure)
    mesh = tf.Mesh(faces, points_transformed)

    # Create VTK polydata for visualization (just a view)
    poly = numpy_to_polydata(points_transformed, faces)

    # Create renderers (left + right viewports, plus text strip at bottom)
    renderer_left = vtk.vtkRenderer()
    renderer_right = vtk.vtkRenderer()
    renderer_text = vtk.vtkRenderer()

    # Viewports: two side-by-side on top (88% height), text bar bottom (12%)
    renderer_left.SetViewport(0.0, 0.12, 0.5, 1.0)
    renderer_right.SetViewport(0.5, 0.12, 1.0, 1.0)
    renderer_text.SetViewport(0.0, 0.0, 1.0, 0.12)
    renderer_text.InteractiveOff()

    # Backgrounds (dark blue/gray matching C++)
    renderer_left.SetBackground(27.0 / 255.0, 43.0 / 255.0, 52.0 / 255.0)
    renderer_right.SetBackground(27.0 / 255.0, 43.0 / 255.0, 52.0 / 255.0)
    renderer_text.SetBackground(0.090, 0.143, 0.173)

    # === LEFT VIEWPORT: Original mesh with curves ===

    # Setup original mesh actor
    mapper_left = vtk.vtkPolyDataMapper()
    mapper_left.SetInputData(poly)
    actor_left = vtk.vtkActor()
    actor_left.SetMapper(mapper_left)
    actor_left.GetProperty().SetColor(0.8, 0.8, 0.8)
    renderer_left.AddActor(actor_left)

    # Create VTK matrix for transformations
    matrix = vtk.vtkMatrix4x4()
    matrix.Identity()
    actor_left.SetUserMatrix(matrix)

    # Wrap in MeshData for interaction
    mesh_data = MeshData(mesh, actor_left, matrix)

    # Setup curve actor on left (initially empty)
    curve_poly = vtk.vtkPolyData()
    curve_poly.Initialize()
    curve_mapper = vtk.vtkPolyDataMapper()
    curve_mapper.SetInputData(curve_poly)
    curve_actor = vtk.vtkActor()
    curve_actor.SetMapper(curve_mapper)

    # Render lines as tubes (GPU-accelerated)
    curve_actor.GetProperty().SetColor(1.0, 0.1, 0.1)  # Red curves
    curve_actor.GetProperty().SetRenderLinesAsTubes(True)
    curve_actor.GetProperty().SetLineWidth(8.0)  # Tube width in pixels

    renderer_left.AddActor(curve_actor)

    # === RIGHT VIEWPORT: Isobands mesh ===

    # Setup isobands mesh actor (initially empty)
    isobands_poly = vtk.vtkPolyData()
    isobands_poly.Initialize()
    mapper_right = vtk.vtkPolyDataMapper()
    mapper_right.SetInputData(isobands_poly)
    actor_right = vtk.vtkActor()
    actor_right.SetMapper(mapper_right)
    actor_right.GetProperty().SetColor(0.8, 0.8, 0.8)
    renderer_right.AddActor(actor_right)

    # === TEXT STRIP ===

    text_time = create_text_actor(
        "Isobands time: 0 ms",
        font_size=38,
        position=(0.03, 0.50),
        justification='left'
    )
    renderer_text.AddViewProp(text_time)

    # === RENDER WINDOW AND INTERACTOR ===

    render_window = vtk.vtkRenderWindow()
    render_window.AddRenderer(renderer_left)
    render_window.AddRenderer(renderer_right)
    render_window.AddRenderer(renderer_text)
    render_window.SetSize(1200, 600)

    interactor = vtk.vtkRenderWindowInteractor()
    interactor.SetRenderWindow(render_window)

    # Setup custom interactor style
    style = IsobandInteractor(mesh_data, isobands_poly, curve_poly, text_time)
    interactor.SetInteractorStyle(style)

    # Share camera between left and right viewports
    renderer_right.SetActiveCamera(renderer_left.GetActiveCamera())

    # Reset cameras
    renderer_left.ResetCamera()
    renderer_right.ResetCamera()

    # Compute initial bands (now that interactor is attached)
    style.compute_bands()

    # Start
    render_window.Render()
    interactor.Start()


if __name__ == "__main__":
    main()

File: cpp/examples/vtk/isobands.cpp
#include "util/vtk_examples.hpp"

#include <trueform/cpp/core/cache.hpp>
#include <trueform/cpp/core/index_type.hpp>
#include <trueform/cpp/core/mesh.hpp>
#include <trueform/cpp/iso/isobands.hpp>
#include <trueform/cpp/spatial/distance.hpp>
#include <trueform/cpp/spatial/primitive.hpp>

#include <vtkActor.h>
#include <vtkInteractorStyleTrackballCamera.h>
#include <vtkMatrix4x4.h>
#include <vtkObjectFactory.h>
#include <vtkPolyData.h>
#include <vtkProperty.h>
#include <vtkRenderWindow.h>
#include <vtkRenderWindowInteractor.h>
#include <vtkRenderer.h>
#include <vtkSmartPointer.h>
#include <vtkTextActor.h>

#include <algorithm>
#include <array>
#include <chrono>
#include <cmath>
#include <cstddef>
#include <cstdint>
#include <iostream>
#include <random>
#include <stdexcept>
#include <string>
#include <vector>

namespace {

namespace examples = trueform::vtk_examples;
/// A field is cut on the mesh the reader gave, and a mixed carrier is what a
/// mesh read from a file is.
using index_t = tf::cpp::default_index_t;
using mixed_storage = tf::polygons_buffer<index_t, float, 3, tf::dynamic_size>;
using mixed_cache = tf::cpp::cache<index_t, float, 3, tf::dynamic_size>;
using mixed_mesh = tf::cpp::mesh<index_t, float, 3, tf::dynamic_size>;

/// What this example holds: the storage the reader gave and the cache that
/// remembers it. `mesh()` is the assembly every entry takes.
struct surface_source {
  mixed_storage polygons;
  mutable mixed_cache cache;

  auto mesh() const -> mixed_mesh {
    return {polygons.faces(), polygons.points(), cache};
  }
};

class isoband_interactor final : public vtkInteractorStyleTrackballCamera {
public:
  static auto New() -> isoband_interactor *;
  vtkTypeMacro(isoband_interactor, vtkInteractorStyleTrackballCamera);

  auto initialize(const surface_source &value, vtkPolyData *bands,
                  vtkPolyData *curves, vtkTextActor *timing) -> void {
    _surface = &value;
    _band_polydata = bands;
    _curve_polydata = curves;
    _timing = timing;
    reset_plane();
  }

  auto compute_bands() -> void {
    constexpr int number_of_levels = 10;
    auto wrapped_offset = std::fmod(_distance, _spacing);
    if (wrapped_offset < 0.0F)
      wrapped_offset += _spacing;

    std::vector<float> cut_values;
    cut_values.reserve(number_of_levels + 3);
    for (int i = -1; i != number_of_levels + 2; ++i)
      cut_values.push_back(_minimum + wrapped_offset +
                           static_cast<float>(i) * _spacing);
    const auto cuts =
        examples::make_array(std::move(cut_values), {number_of_levels + 3});

    const auto period =
        static_cast<long long>(std::floor(_distance / _spacing));
    const auto parity = static_cast<int>((period % 2 + 2) % 2);
    std::vector<std::int32_t> selected_values;
    for (int band = 0; band != number_of_levels + 2; ++band)
      if ((band & 1) == parity)
        selected_values.push_back(static_cast<std::int32_t>(band));
    const auto selected =
        examples::make_array(std::move(selected_values),
                             {static_cast<int>((number_of_levels + 2) / 2)});

    const auto start = std::chrono::steady_clock::now();
    const auto result = tf::cpp::isobands_with_curves_selected(
        _surface->mesh(), _scalars, cuts, selected);
    _times.add(examples::elapsed_seconds(start));
    const auto label = std::string("Isobands time: ") +
                       examples::format_milliseconds(_times.average());
    _timing->SetInput(label.c_str());

    const auto band_polydata = examples::to_vtk_polydata(result.mesh);
    examples::replace_polydata(_band_polydata, band_polydata);
    const auto curve_polydata = examples::to_vtk_polydata(result.curves);
    examples::replace_polydata(_curve_polydata, curve_polydata);
    if (this->Interactor)
      this->Interactor->Render();
  }

  auto OnKeyPress() -> void override {
    if (std::string(this->Interactor->GetKeySym()) == "n") {
      randomize_plane();
      compute_bands();
      return;
    }
    vtkInteractorStyleTrackballCamera::OnKeyPress();
  }

  auto OnMouseWheelForward() -> void override {
    if (this->Interactor->GetControlKey()) {
      _distance += _spacing * 0.1F;
      compute_bands();
      return;
    }
    vtkInteractorStyleTrackballCamera::OnMouseWheelForward();
  }

  auto OnMouseWheelBackward() -> void override {
    if (this->Interactor->GetControlKey()) {
      _distance -= _spacing * 0.1F;
      compute_bands();
      return;
    }
    vtkInteractorStyleTrackballCamera::OnMouseWheelBackward();
  }

protected:
  isoband_interactor() = default;
  ~isoband_interactor() override = default;

private:
  auto reset_plane() -> void {
    if (!_surface || _surface->polygons.points_buffer().size() == 0)
      throw std::invalid_argument("isobands requires a non-empty mesh");
    auto normal = examples::point3<float>{1.0F, 2.0F, 1.0F};
    const auto length = std::sqrt(6.0F);
    for (auto &coordinate : normal)
      coordinate /= length;
    set_plane(examples::point_mean(_surface->polygons.points_buffer()), normal);
  }

  auto randomize_plane() -> void {
    std::normal_distribution<float> normal_distribution(0.0F, 1.0F);
    examples::point3<float> normal{normal_distribution(_generator),
                                   normal_distribution(_generator),
                                   normal_distribution(_generator)};
    const auto normal_length = std::sqrt(
        normal[0] * normal[0] + normal[1] * normal[1] + normal[2] * normal[2]);
    for (auto &coordinate : normal)
      coordinate /= normal_length;

    const auto points = _surface->polygons.points();
    std::uniform_int_distribution<std::size_t> point_distribution(
        0, points.size() - 1);
    const auto source = points[point_distribution(_generator)];
    set_plane({source[0], source[1], source[2]}, normal);
  }

  auto set_plane(const examples::point3<float> &point,
                 const examples::point3<float> &normal) -> void {
    const auto plane_d =
        -(normal[0] * point[0] + normal[1] * point[1] + normal[2] * point[2]);
    const auto plane = tf::cpp::primitive<float>(
        tf::cpp::primitive_kind::plane,
        examples::make_array<float>({normal[0], normal[1], normal[2], plane_d},
                                    {4}));
    const auto points = tf::cpp::primitive<float>(
        tf::cpp::primitive_kind::point,
        examples::points_array(_surface->polygons.points_buffer()));
    _scalars = tf::cpp::distance(points, plane).batch();
    _distance = 0.0F;
    const auto limits = std::minmax_element(_scalars.begin(), _scalars.end());
    _minimum = *limits.first;
    _maximum = *limits.second;
    _spacing = (_maximum - _minimum) / 10.0F;
  }

  const surface_source *_surface = nullptr;
  vtkPolyData *_band_polydata = nullptr;
  vtkPolyData *_curve_polydata = nullptr;
  vtkTextActor *_timing = nullptr;
  tf::cpp::nd_array<float> _scalars;
  float _distance = 0.0F;
  float _minimum = 0.0F;
  float _maximum = 0.0F;
  float _spacing = 0.1F;
  examples::rolling_average _times{100};
  std::mt19937 _generator{std::random_device{}()};
};

vtkStandardNewMacro(isoband_interactor);

} // namespace

int main(int argc, char **argv) {
  try {
    const auto help = argc == 2 && (std::string(argv[1]) == "-h" ||
                                    std::string(argv[1]) == "--help");
    if (argc != 2 || help) {
      auto &stream = help ? std::cout : std::cerr;
      stream << "Usage: " << argv[0] << " mesh.stl\n\n"
             << "Controls:\n"
             << "  N              Randomize cutting plane\n"
             << "  Ctrl + Scroll  Move isoband levels\n"
             << "  Mouse drag     Rotate camera\n";
      return help ? 0 : 2;
    }

    surface_source surface{examples::as_mixed_mesh(tf::cpp::read_stl(argv[1])),
                           {}};
    examples::materialize_center_and_scale(surface.polygons, 10.0F);
    const auto source_polydata = examples::to_vtk_polydata(surface.polygons);

    auto renderer = vtkSmartPointer<vtkRenderer>::New();
    const auto text_renderer = examples::make_text_strip(renderer);

    const auto mesh_actor = examples::make_actor(source_polydata);
    mesh_actor->GetProperty()->SetColor(0.5, 0.5, 0.55);
    mesh_actor->GetProperty()->SetOpacity(0.25);
    auto identity = vtkSmartPointer<vtkMatrix4x4>::New();
    identity->Identity();
    mesh_actor->SetUserMatrix(identity);
    renderer->AddActor(mesh_actor);

    auto band_polydata = vtkSmartPointer<vtkPolyData>::New();
    band_polydata->Initialize();
    const auto band_actor = examples::make_actor(band_polydata);
    band_actor->GetProperty()->SetColor(0.0, 0.659, 0.604);
    renderer->AddActor(band_actor);

    auto curve_polydata = vtkSmartPointer<vtkPolyData>::New();
    curve_polydata->Initialize();
    const auto curve_actor = examples::make_actor(curve_polydata);
    curve_actor->GetProperty()->SetColor(0.0, 0.835, 0.745);
    curve_actor->GetProperty()->SetRenderLinesAsTubes(true);
    curve_actor->GetProperty()->SetLineWidth(8.0);
    renderer->AddActor(curve_actor);

    const auto timing =
        examples::make_text_actor("Isobands time: 0 ms", 38, {0.03, 0.5});
    text_renderer->AddViewProp(timing);

    auto window = vtkSmartPointer<vtkRenderWindow>::New();
    window->AddRenderer(renderer);
    window->AddRenderer(text_renderer);
    window->SetSize(800, 600);

    auto interactor = vtkSmartPointer<vtkRenderWindowInteractor>::New();
    interactor->SetRenderWindow(window);
    auto style = vtkSmartPointer<isoband_interactor>::New();
    style->initialize(surface, band_polydata, curve_polydata, timing);
    interactor->SetInteractorStyle(style);

    style->compute_bands();
    window->Render();
    interactor->Start();
    return 0;
  } catch (const std::exception &error) {
    std::cerr << "isobands: " << error.what() << '\n';
    return 1;
  }
}
